inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/dataprovider/generic/BoundDataProvider.java

```diff
@@ -0,0 +1,163 @@
+
+    /**
+     * The data to which the value is assigned
+     */
+    protected final String dataKey;
+
+    /**
+     * The name of the data provider, only used to provide a meaningful name via getName()
+     */
+    private final String providerName;
+
+    /**
+     * Reference to the class of teh data provider.
+     * This reference is held to prevent the {@link rocks.inspectit.oce.core.instrumentation.injection.ClassInjector} from reusing this provider.
+     */
+    private final InjectedClass<?> providerClass;
+
+    /**
+     * Reference to the actual provider instance.
+     * This corresponds to the value of {@link #providerClass} for {@link GenericDataProviderTemplate#INSTANCE}.
+     */
+    protected final IGenericDataProvider provider;
+
+
+    protected BoundDataProvider(String dataKey, ResolvedGenericDataProviderConfig providerConfig, InjectedClass<?> providerClass) {
+        this.dataKey = dataKey;
+        providerName = providerConfig.getName();
+        this.providerClass = providerClass;
+        try {
+            provider = (IGenericDataProvider) providerClass.getInjectedClassObject().get().getField("INSTANCE").get(null);
+        } catch (Exception e) {
+            throw new IllegalArgumentException("The given provider is not based on the GenericDatProviderTemplate");
+        }
+    }
+
+    @Override
+    public String getName() {
+        return "DataProvider '" + providerName + "' for '" + dataKey + "'";
+    }
+
+    /**
+     * Binds a data provider to the given input argument values.
+     *
+     * @param dataKey             the data key udner which the result of this provider will be stored
+     * @param providerConfig      the configuration of the used data provider
+     * @param provider            the generated provider class
+     * @param constantAssignments a map mapping input variable names to their constant values
+     * @param dynamicAssignments  a map mapping input variables to a function which is used to derive
+     *                            the parameter value when the provider is invoked
+     * @return
+     */
+    public static BoundDataProvider bind(String dataKey,
+                                         ResolvedGenericDataProviderConfig providerConfig,
+                                         InjectedClass<?> provider,
+                                         Map<String, Object> constantAssignments,
+                                         Map<String, Function<ExecutionContext, Object>> dynamicAssignments) {
+
+        if (dynamicAssignments.isEmpty()) {
+            return new ConstantOnlyBoundDataProvider(dataKey, providerConfig, provider, constantAssignments);
+        } else {
+            return new DynamicBoundDataProvider(dataKey, providerConfig, provider, constantAssignments, dynamicAssignments);
+        }
+    }
+}
+
+class ConstantOnlyBoundDataProvider extends BoundDataProvider {
+
+    private Object[] arguments;
```

Comment:
this should also be final, as well as the two fields in the dynamic data provider.